test/unit/permutations/test_permutations_distributed.cpp

```diff
@@ -33,86 +41,136 @@ struct PermutationsDistTestMC : public TestWithCommGrids {};
+
+  std::vector<SizeType> perms(to_sizet(i_end - i_begin));
+  std::generate(perms.rbegin(), perms.rend(), [n = 0]() mutable { return n++; });
+
+  return {m, mb, i_begin_tile, i_last_tile, perms};
+}
+
+// Helper that just checks that given geometry and permutations are compatible.
+testcase_t customPermutation(SizeType m, SizeType mb, SizeType i_begin_tile, SizeType i_last_tile,
+                             std::vector<SizeType> perms) {
+  const auto [i_begin, i_end] = tileToElementRange(m, mb, i_begin_tile, i_last_tile + 1);
 
+  const std::size_t nperms = to_sizet(i_end - i_begin);
+  DLAF_ASSERT(perms.size() == nperms, perms.size(), nperms);
+
+  return {m, mb, i_begin_tile, i_last_tile, std::move(perms)};
+}
+
+const std::vector<testcase_t> params = {
     // simple setup for a (3, 2) process grid,
-    {6, 2, 0, 2},
+    mirrorPermutation(6, 2, 0, 2),
+    customPermutation(6, 2, 0, 2, {2, 0, 1, 4, 5, 3}),
     // entire range of tiles is inculded
-    {10, 3, 0, 3},
-    {17, 5, 0, 3},
+    mirrorPermutation(10, 3, 0, 3),
+    customPermutation(10, 3, 0, 3, {0, 2, 3, 4, 6, 8, 1, 5, 7, 9}),
+    customPermutation(10, 3, 0, 3, {8, 9, 3, 5, 2, 7, 1, 4, 0, 6}),
+    mirrorPermutation(17, 5, 0, 3),
     // only a subset of processes participate
-    {10, 3, 1, 2},
+    mirrorPermutation(10, 3, 1, 2),
     // a single tile matrix
-    {10, 10, 0, 0},
+    mirrorPermutation(10, 10, 0, 0),
     // each process has multiple tiles
-    {31, 6, 1, 3},
-    {50, 4, 1, 8},
+    mirrorPermutation(31, 6, 1, 3),
+    mirrorPermutation(50, 4, 1, 8),
 };
 
 template <class T, Device D, Coord C>
 void testDistPermutations(comm::CommunicatorGrid grid, SizeType n, SizeType nb, SizeType i_begin,
-                          SizeType i_end) {
+                          SizeType i_last, std::vector<SizeType> perms) {
   const GlobalElementSize size(n, n);
   const TileElementSize block_size(nb, nb);
-  Index2D src_rank_index(std::max(0, grid.size().rows() - 1), std::min(1, grid.size().cols() - 1));
+  const Index2D src_rank_index(std::max(0, grid.size().rows() - 1), std::min(1, grid.size().cols() - 1));
 
-  Distribution dist(size, block_size, grid.size(), grid.rank(), src_rank_index);
-  Matrix<SizeType, Device::CPU> perms_h(LocalElementSize(n, 1), TileElementSize(nb, 1));
-  Matrix<T, Device::CPU> mat_in_h(dist);
-  Matrix<T, Device::CPU> mat_out_h(dist);
+  const Distribution dist(size, block_size, grid.size(), grid.rank(), src_rank_index);
 
-  SizeType index_start = dist.globalElementFromGlobalTileAndTileElement<C>(i_begin, 0);
-  SizeType index_finish = dist.globalElementFromGlobalTileAndTileElement<C>(i_end, 0) +
-                          dist.tileSize(GlobalTileIndex(i_end, i_end)).get<C>();
-  dlaf::matrix::util::set(perms_h, [index_start, index_finish](GlobalElementIndex i) {
-    if (index_start > i.row() || i.row() >= index_finish)
-      return SizeType(0);
+  const auto [index_start, index_end] = tileToElementRange(n, nb, i_begin, i_last + 1);
```

Comment:
note: clang@12 complains because structured binding does not create variables but just "names", and it conflicts with what can be captured in lambdas (just variables).